nova_vm/src/engine/bytecode/vm.rs

```diff
@@ -927,13 +992,62 @@ impl Vm {
                                 _ => unreachable!(),
                             });
                         } else {
-                            vm.iterator_stack.pop();
                             vm.result = None;
                         }
                     }
+                    VmIterator::ArrayValues(iter) => {
+                        let result = iter.next(agent);
+                        if result.is_err() {
+                            vm.iterator_stack.pop();
+                            result?;
+                        }
+                        let result = result.unwrap();
+                        vm.result = result;
+                    }
+                    VmIterator::GenericIterator(iter) => {
+                        let result =
+                            call(agent, iter.next_method, iter.iterator.into_value(), None);
+                        if result.is_err() {
+                            vm.iterator_stack.pop();
+                            result?;
+                        }
+                        let result = result.unwrap();
+                        if !result.is_object() {
+                            return Err(agent.throw_exception(
+                                ExceptionType::TypeError,
+                                "Iterator returned a non-object result",
+                            ));
+                        }
+                        vm.result = Some(result);
+                    }
+                }
+            }
+            Instruction::IteratorValue => {
+                if matches!(
+                    vm.iterator_stack.last().as_ref().unwrap(),
+                    VmIterator::GenericIterator(_)
+                ) {
+                    // Generic iterators have to access the "value" property of the result object.
```

Comment:
**nitpick:** Wouldn't it be better to instead have a `VmIterator::next` method that abstracts away the differences between implementations and returns a `JsResult<Option<Value>>`? IIUC, this would simplify things, and would avoid the need for the `IteratorValue` instruction, right?

Reply:
This was an excellent idea! I refactored to join these together into the `IteratorStepValue` spec algorithm, essentially.

It was already doing that, only that the `IteratorStepValue` also says to set the `done` value to `true` everywhere. We're still skipping that since the iterator record is about to be destroyed.